src/Eshopworld.Messaging/Messenger.cs

```diff
@@ -97,14 +110,26 @@ public void Receive<T>(Action<T> callback, int batchSize = 10)
         public async Task Subscribe<T>(Action<T> callback, string subscriptionName, int batchSize = 10)
             where T : class
         {
-            if (!MessageSubs.TryAdd(typeof(T), MessagesIn.OfType<T>().Subscribe(callback)))
+            if (!MessageSubs.TryAdd(GetTypeName<T>(), MessagesIn.OfType<T>().Subscribe(callback)))
             {
                 throw new InvalidOperationException("You already added a callback to this message type. Only one callback per type is supported.");
             }
 
             await ((TopicAdapter<T>)SetupMessageType<T>(batchSize, MessagingTransport.Topic)).StartReading(subscriptionName).ConfigureAwait(false);
         }
 
+        /// <inheritdoc />
+        public async Task Subscribe<T>(Action<T> callback, string subscriptionName, string topicName, int batchSize = 10) where T : class
+        {
+            CheckTopicName(topicName);
+            if (!MessageSubs.TryAdd(topicName, MessagesIn.OfType<T>().Subscribe(callback)))
+            {
+                throw new InvalidOperationException("You already added a callback to this topic. Only one callback per topic is supported.");
+            }
+
+            await ((TopicAdapter<T>)SetupMessageType<T>(batchSize, MessagingTransport.Topic, topicName)).StartReading(subscriptionName).ConfigureAwait(false);
+        }
+
         /// <inheritdoc />
         public void CancelReceive<T>()

```

Comment:
Do we also need a way to unsubscribe from the topic? Not sure about the logic/reasons for this method but if it exists for auto-derived topics, it should also for manually provided?

Reply:
it makes sense

Reply:
great, I will extend the `CancelReceive` method to accept topic name as a parameter

Reply:
I will also add `Lock`, `Complete`, `Abandon`,  `Error` and `SetBatchSize` from `IMessageOperations` as they all need the topic name

Reply:
Update here:
https://github.com/eShopWorld/core/pull/31

Please guys review when you can

Reply:
updated CC @C-M-Ryan @chomick